### PR TITLE
Added custom PCF tags on containers

### PR DESCRIPTION
**Pull request by NouemanKHAL**
### What does this PR do?

Adds custom tags on PCF containers from environment variables based on an include and exclude lists mechanism.

### Motivation

Feature request

### Describe your test plan

Unit tests and live test on a PCF environment

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -532,6 +532,8 @@ func InitConfig(config Config) {
 	config.BindEnvAndSetDefault("cloud_foundry_bbs.ca_file", "")
 	config.BindEnvAndSetDefault("cloud_foundry_bbs.cert_file", "")
 	config.BindEnvAndSetDefault("cloud_foundry_bbs.key_file", "")
+	config.BindEnvAndSetDefault("cloud_foundry_bbs.env_include", []string{})
```
Comment (zippolyte) [inline pkg/config/config.go:535]: We should also document those in the `config_template.yaml` file

File: pkg/util/cloudfoundry/types.go [hunk 2]
```diff
@@ -270,10 +279,38 @@ func (dlrp *DesiredLRP) GetTagsFromDLRP() []string {
 			tags = append(tags, fmt.Sprintf("%s:%s", k, v))
 		}
 	}
+	tags = append(tags, dlrp.TagsFromEnv...)
 	sort.Strings(tags)
 	return tags
 }
 
+func isAllowedTag(value string, includeList, excludeList []*regexp.Regexp) bool {
```
Comment (zippolyte) [inline pkg/util/cloudfoundry/types.go:287]: Let's also port the unit test of this function from the nozzle to here

File: pkg/util/cloudfoundry/types_test.go [hunk 3]
```diff
@@ -147,6 +156,16 @@ var BBSModelD1 = models.DesiredLRP{
...
+	EnvironmentVariables: []*models.EnvironmentVariable{
+		{
+			Name:  "we ignore this",
+			Value: "some value",
+		},
+		{
+			Name:  "AD_DATADOGHQ_COM",
+			Value: "{\"xxx\": {}}", // make this a valid JSON
+		},
+	},
```
Comment (zippolyte) [inline pkg/util/cloudfoundry/types_test.go:168]: This isn't needed

**Review by zippolyte — CHANGES_REQUESTED** [review]

File: go.sum [hunk 4]
```diff
@@ -116,12 +116,16 @@ github.com/DataDog/datadog-go v4.4.0+incompatible h1:R7WqXWP4fIOAqWJtUKmSfuc7eDs
 github.com/DataDog/datadog-go v4.4.0+incompatible/go.mod h1:LButxg5PwREeZtORoXG3tL4fMGNddJ+vMq1mwgfaqoQ=
```
Comment (zippolyte) [inline go.sum:116]: There's still some go.sum changes

File: releasenotes/notes/pcf-containers-tag-cfbc2f0441f3b5ef.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    PCF Containers custom tags can be extracted from environment variables based on a include based on an include and exclude lists mechanism.
```
Comment (zippolyte) [inline releasenotes/notes/pcf-containers-tag-cfbc2f0441f3b5ef.yaml:11]:
```suggestion
    PCF Containers custom tags can be extracted from environment variables based on an include and exclude lists mechanism.
```

File: pkg/config/config.go (hunk 1 above)
Comment (NouemanKHAL) [reply pkg/config/config.go:535]: Done!
Let me know if it needs to be modified.

File: pkg/config/config_template.yaml [hunk 6]
```diff
@@ -1960,6 +1960,16 @@ api_key:
...
 
+  ## @param env_include - List of strings - optional - default: ""
+  ## List of regex expressions to allow a set of environment variables to be included as container tags
+  #
+  # env_include: ""
+
+  ## @param env_exclude - List of strings - optional - default: ""
+  ## List of regex expressions to forbid a set of environment variables to be included as container tags
+  #
+  # env_exclude: ""
```
Comment (zippolyte) [inline pkg/config/config_template.yaml:1971]:
```suggestion
  ## @param env_include - List of strings - optional - default: []
  ## List of regular expressions to allow a set of environment variables to be included as container tags
  #
  # env_include: []

  ## @param env_exclude - List of strings - optional - default: []
  ## List of regular expressions to forbid a set of environment variables to be included as container tags
  #
  # env_exclude: []
```

File: releasenotes/notes/pcf-containers-tag-cfbc2f0441f3b5ef.yaml [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    PCF Containers custom tags can be extracted from environment 
+    variables based on an include based on an include and exclude lists mechanism.
```
Comment (zippolyte) [inline releasenotes/notes/pcf-containers-tag-cfbc2f0441f3b5ef.yaml:12]:
```suggestion
    PCF Containers custom tags can be extracted from environment 
    variables based on an include and exclude lists mechanism.
```

File: pkg/util/cloudfoundry/types_test.go [hunk 8]
```diff
@@ -238,3 +279,32 @@ func TestGetVcapServicesMap(t *testing.T) {
...
+
+	// reject strings in the excludeList
+	includeList = []*regexp.Regexp{}
+	excludeList = []*regexp.Regexp{regexp.MustCompile("exclude.*")}
+
+	result = isAllowedTag("aRandomValue", includeList, excludeList)
+	assert.EqualValues(t, true, result)
+
+	result = isAllowedTag("excludeTag", includeList, excludeList)
+	assert.EqualValues(t, false, result)
```
Comment (zippolyte) [inline pkg/util/cloudfoundry/types_test.go:309]: let's just add a bit more complex case, where there is an intersection between whitelist and blacklist, to make sure the blacklist precedence is respected

File: releasenotes/notes/pcf-containers-tag-cfbc2f0441f3b5ef.yaml (hunk 7 above)
Comment (zippolyte) [reply releasenotes/notes/pcf-containers-tag-cfbc2f0441f3b5ef.yaml:12]: this is still not solved

File: pkg/util/cloudfoundry/types_test.go [hunk 9]
```diff
@@ -238,3 +279,39 @@ func TestGetVcapServicesMap(t *testing.T) {
...
+
+	// reject strings in the excludeList even if they exist in the includeList
+	includeList = []*regexp.Regexp{regexp.MustCompile("include.*")}
+	excludeList = []*regexp.Regexp{regexp.MustCompile("exclude.*"), regexp.MustCompile("include.*")}
+
+	result = isAllowedTag("includeTag", includeList, excludeList)
+	assert.EqualValues(t, false, result)
```
Comment (zippolyte) [inline pkg/util/cloudfoundry/types_test.go:316]: Just to be more precise
```suggestion
	includeList = []*regexp.Regexp{regexp.MustCompile("include.*")}
	excludeList = []*regexp.Regexp{regexp.MustCompile("includeExclude.*")}

	result = isAllowedTag("includeTag", includeList, excludeList)
	assert.EqualValues(t, true, result)
	result = isAllowedTag("includeExcludeTag", includeList, excludeList)
	assert.EqualValues(t, false, result)
```

**Review by zippolyte — APPROVED** [review]

**Review by olivielpeau — APPROVED** [review]
Approving for agent core (only relevant file is `config.go`)

**Review by sarina-dd — APPROVED** [review]

**Pull request merged**